# 🙅🏽‍♂️ WITHOUT PARAMETER AND WITHOUT RETURN VALUE

In [57]:
def greet():
    print("Hello world !")
    

In [58]:
greet()

Hello world !


# 💬 WITH PARAMETER

In [59]:
def greet(name):
    print(f"Hello ! {name}")

In [60]:
#greet()

In [61]:
greet("Ujwal Agrawal")

Hello ! Ujwal Agrawal


# 🤑WITH RETURN VALUE

In [62]:
def SEP_EVEN_ODD(num_list):
    EVEN=[]
    ODD=[]
    for num in num_list:
        if num%2==0:
            EVEN.append(num)
        else:
            ODD.append(num)
    print(f"EVEN_NUMBER : {EVEN}")
    print(f"ODD_NUMBER :{ODD}")             

In [63]:

SEP_EVEN_ODD([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18])

EVEN_NUMBER : [2, 4, 6, 8, 10, 12, 14, 16, 18]
ODD_NUMBER :[1, 3, 5, 7, 9, 11, 13, 15, 17]


# WITH DEFAULT PARAMETERS

In [64]:
def GREET(name, greeting="have a great day.."):
    print(f"Hello! {name} {greeting}")

In [65]:
GREET("Ujwal")

Hello! Ujwal have a great day..


# 🔁 RECURSIVE FUNCTION

In [66]:
def countdown(n):
    import time
    if n<1:
        return 1
    print(n)
    time.sleep(1)
    countdown(n-1)

In [67]:
countdown(10)

10


9
8
7
6
5
4
3
2
1


In [68]:
def factorial(n):# n=5
    if n<=1:
        return 1
    return n*factorial(n-1)# here factorial(n-1) will be 4 from recursive function


In [69]:
factorial(5)

120

---

### 🌟 What is a **Factorial**?

Factorial means **multiply all numbers from `n` down to 1**.
We write it as `n!` (say: “n factorial”).

🧮 Example:
**5! = 5 × 4 × 3 × 2 × 1 = 120**

💡 Quick facts:

* **0! = 1** (just remember it!)
* Used in **math, probability, and counting things**!

Sure! Here's a super simple explanation with just the idea and output:

---

Factorial gives the total when you multiply a number by all smaller numbers down to 1.


📌 **3! means "factorial of 3" → answer is 6**

📌 **4! → answer is 24**

📌 **5! → answer is 120**

📌 **0! → answer is 1**



---


# 📖 DATA VALIDATION FUNCTION 
##### VALIDATE USER INPUT

In [70]:
def is_positive():
    num=int(input("enter number"))
    if num > 1 :
        print(f"{num} is Positive ➕")
    else :
        print(f"{num} is Negative ➖")    

In [71]:
is_positive()

6 is Positive ➕


# 🔖WORKING WITH LIST

In [72]:
def find_largest(NUM_LIST):
    LARGEST_NUM=0
    for num in NUM_LIST:
        if num > LARGEST_NUM:
            LARGEST_NUM=num
    return LARGEST_NUM        

In [73]:
find_largest([1,2,3,4,5])

5

# 🪹NESTED FUNCTION
##### USE HELPER FUNCTION INSIDE ANOTHER FUNCTION


---

### 📝 **Practice Task:**

**Create a function** called `check_even_odd` that:

#### ✅ Conditions:

1. Takes **one number** as input.
2. Checks if the number is **even** or **odd**.
3. **Prints** `"Even number"` if even, or `"Odd number"` if odd.
4. Do **not use return**, just use `print()`.

---


In [74]:
def CHECK_EVEN_ODD(NUM):
    def CHECK_EVEN(NUM):
        if NUM%2==0:
            print("even")
        else:
            CHECK_ODD(NUM)
    def CHECK_ODD(NUM):
        if NUM% 2!=0:
            print("odd")
    CHECK_EVEN(NUM)        

In [75]:
CHECK_EVEN_ODD(5)

odd


# ⁉️HANDING EXCEPTIONS

---

### 🌟 **Error Handling in Python: `try` and `except`** 🌟

Imagine you're walking through a **maze of code** 🏃‍♂️, but **oops!** — sometimes, **errors** pop up like obstacles! 😱

#### 🛡️ **The Superpower of `try` and `except`** 🛡️

Instead of **stopping your journey** when you face an obstacle, you simply **handle it and keep going**! 💪

---

### ✅ **How It Works:**

1. **`try:`** 🧭
   This is where you **give it a go**. It holds the **risky code** that might cause an error.

2. **`except:`** 💥
   If anything goes wrong, the **`except` block** steps in to **save the day**! You can handle the error in a controlled way instead of crashing. 🚀
   except: This block catches the error if one occurs and allows you to handle it.

---

### 💡 **Why Use `try` and `except`?**

* **Error-Free Zone** 🛡️: The program keeps going, even if it stumbles!
* **Friendly Fixes** 💬: Catch and handle errors with **helpful messages**. ✨

---

### 🚀 **Keep Going, No Stopping!** 💡

With `try` and `except`, your program is like a **superhero** 🦸‍♂️, always bouncing back from errors! ✨


In [76]:
def Divide(a,b):
    return a/b

In [78]:
#Divide(2,0)

In [79]:
def Division(a,b):
    try:
        return a/b
    except ZeroDivisionError :
        return "Divsion by zero is not allowed "
    

In [80]:
Division(2,0)

'Divsion by zero is not allowed '

# 💁🏽‍♀️FUNCTION WITH CONDITION
##### IF ELIF ELSE
📝 Condition:
Create a function that:
- Takes an integer as input.
- Prints "Positive" if the number is greater than 0.
- Prints "Zero" if the number is exactly 0.
- Prints "Negative" if the number is less than 0.

In [82]:
def CHECK_NUM(NUM):
    if NUM>0:
        print(f"{NUM} is positive")
    elif NUM==0:
        print(f"{NUM} is zero")
    else :
        print(f"{NUM} is negative")        

In [83]:
CHECK_NUM(5)

5 is positive
